common/src/main/scala/ClassTwoAndThreeResult.scala

```diff
@@ -18,32 +18,74 @@ package eoi
 
 import spire.math.Interval
 import cats.syntax.applicative._
-import cats.syntax.traverse._
-import cats.instances.list._
 import cats.syntax.apply._
 import spire.implicits._
 import java.time.LocalDate
 
+trait ClassTwoOrThree {
+  def noOfWeeks: Int
+  def rate: BigDecimal
+  def lowerEarningLimit: Explained[BigDecimal]
+  def qualifyingEarningsFactor: Explained[BigDecimal]
+  def finalDate: Option[LocalDate]
+}
+
+case class ClassTwo(
+  weeklyRate: ClassTwoRates,
+  smallEarningsException: Option[BigDecimal],
+  hrpDate: Option[LocalDate],
+  finalDate: Option[LocalDate],
+  noOfWeeks: Int = 52,
+  qualifyingRate: BigDecimal
+) extends ClassTwoOrThree {
+  def rate = weeklyRate.default
+  def lowerEarningLimit = ((qualifyingRate + 50) / noOfWeeks) gives
+    s"lowerEarningLimit: (qualifyingRate + 50) / noOfWeeks = ($qualifyingRate + 50) / $noOfWeeks"
+
+  def qualifyingEarningsFactor: Explained[BigDecimal] = qualifyingRate.pure[Explained]
+}
+
+case class ClassThree(
+  finalDate: Option[LocalDate],
+  weekRate: BigDecimal,
+  noOfWeeks: Int = 52,
+  lel: BigDecimal,
+  qualifyingRate: Option[BigDecimal]
+) extends ClassTwoOrThree {
+  def rate = weekRate
+  def qualifyingEarningsFactor: Explained[BigDecimal] = qualifyingRate match {
+    case Some(r) => r.pure[Explained]
+    case None =>
+      (lel * noOfWeeks - 50) gives s"qualifyingRate: lel * noOfWeeks - 50 = $lel * $noOfWeeks - 50"
+  }
+    
+  def lowerEarningLimit = lel.pure[Explained]
+}
+
+
 case class ClassTwoAndThreeResult[A <: ClassTwoOrThree](
   on: LocalDate,
```

Comment:
`on` looks like a tax year in general rather than a specific date. The logic for `finalDate` requires that this date is always 6th April which seems a bit magic - if 6th April is passed is not passed in `finalDate` will be wrong

Reply:
That's a fair point. I did consider using `TaxYear`, but figured that could interfere with the partial years (which we still don't know how we're meant to implement and are not supported in the access app anyway). 

What about if I get the calculate method to pass in always the lower bound and then make the constructor protected?

Reply:
oh yeah, we do need a way to handle 1985. Depending on what the rules are for 1985, doing tax year plus seven years minus one day might not work there then - but we don't know the rules for 1985 yet so it's difficult to say how that part of the code should be adapted.

Reply:
In that case I think we'd give an explicit final date.